collections/quad.go

```diff
@@ -0,0 +1,425 @@
+		return *t.k1
+	}
+	return x
+}
+
+// K2 returns the second part of the key. If nil, the zero value is returned.
+func (t Quad[K1, K2, K3, K4]) K2() (x K2) {
+	if t.k2 != nil {
+		return *t.k2
+	}
+	return x
+}
+
+// K3 returns the third part of the key. If nil, the zero value is returned.
+func (t Quad[K1, K2, K3, K4]) K3() (x K3) {
+	if t.k3 != nil {
+		return *t.k3
+	}
+	return x
+}
+
+// K4 returns the fourth part of the key. If nil, the zero value is returned.
+func (t Quad[K1, K2, K3, K4]) K4() (x K4) {
+	if t.k4 != nil {
+		return *t.k4
+	}
+	return x
+}
+
+// QuadPrefix creates a new Quad instance composed only of the first part of the key.
+func QuadPrefix[K1, K2, K3, K4 any](k1 K1) Quad[K1, K2, K3, K4] {
+	return Quad[K1, K2, K3, K4]{k1: &k1}
+}
+
+// QuadSuperPrefix creates a new Quad instance composed only of the first two parts of the key.
+func QuadSuperPrefix[K1, K2, K3, K4 any](k1 K1, k2 K2) Quad[K1, K2, K3, K4] {
+	return Quad[K1, K2, K3, K4]{k1: &k1, k2: &k2}
+}
+
+// QuadSuperPrefix3 creates a new Quad instance composed only of the first three parts of the key.
+func QuadSuperPrefix3[K1, K2, K3, K4 any](k1 K1, k2 K2, k3 K3) Quad[K1, K2, K3, K4] {
+	return Quad[K1, K2, K3, K4]{k1: &k1, k2: &k2, k3: &k3}
+}
+
+// QuadKeyCodec instantiates a new KeyCodec instance that can encode the Quad, given
+// the KeyCodecs of the four parts of the key, in order.
+func QuadKeyCodec[K1, K2, K3, K4 any](keyCodec1 codec.KeyCodec[K1], keyCodec2 codec.KeyCodec[K2], keyCodec3 codec.KeyCodec[K3], keyCodec4 codec.KeyCodec[K4]) codec.KeyCodec[Quad[K1, K2, K3, K4]] {
+	return quadKeyCodec[K1, K2, K3, K4]{
+		keyCodec1: keyCodec1,
+		keyCodec2: keyCodec2,
+		keyCodec3: keyCodec3,
+		keyCodec4: keyCodec4,
+	}
+}
+
+// NamedQuadKeyCodec instantiates a new KeyCodec instance that can encode the Quad, given
+// the KeyCodecs of the four parts of the key, in order.
+// The provided names are used to identify the parts of the key in the schema for indexing.
+func NamedQuadKeyCodec[K1, K2, K3, K4 any](key1Name string, keyCodec1 codec.KeyCodec[K1], key2Name string, keyCodec2 codec.KeyCodec[K2], key3Name string, keyCodec3 codec.KeyCodec[K3], key4Name string, keyCodec4 codec.KeyCodec[K4]) codec.KeyCodec[Quad[K1, K2, K3, K4]] {
+	return quadKeyCodec[K1, K2, K3, K4]{
+		name1:     key1Name,
+		keyCodec1: keyCodec1,
+		name2:     key2Name,
+		keyCodec2: keyCodec2,
+		name3:     key3Name,
+		keyCodec3: keyCodec3,
+		name4:     key4Name,
+		keyCodec4: keyCodec4,
+	}
+}
+
+type quadKeyCodec[K1, K2, K3, K4 any] struct {
+	name1, name2, name3, name4 string
+	keyCodec1                  codec.KeyCodec[K1]
+	keyCodec2                  codec.KeyCodec[K2]
+	keyCodec3                  codec.KeyCodec[K3]
+	keyCodec4                  codec.KeyCodec[K4]
+}
+
+type jsonQuadKey [4]json.RawMessage
+
+// EncodeJSON encodes Quads to json
+func (t quadKeyCodec[K1, K2, K3, K4]) EncodeJSON(value Quad[K1, K2, K3, K4]) ([]byte, error) {
+	json1, err := t.keyCodec1.EncodeJSON(*value.k1)
```

Comment:
`value.kn` might be `nil`, I would recommend to either check if before or to use a safe function to get the value out of the pointer (see previous comment).